Split up IPC messages

clumens opened the pull request:
There are currently cts-lab failures with the Reattach and Remote* tests caused by this PR that I am still in the middle of tracking down.

clumens commented on lib/common/ipc_server.c:
```diff
@@ -711,24 +711,41 @@ pcmk__ipc_send_iov(pcmk__client_t *c, struct iovec *iov, uint32_t flags)
+
         if (flags & crm_ipc_server_free) {
             pcmk_free_ipc_event(iov);
         }
```
I've found this patch useful for debugging purposes, but we don't necessarily have to keep it.  One annoyance is that our logging cuts off after a certain number of characters, so very long XML messages won't get logged in their entirety anyway.  Still, it's usually enough to see just the beginning to know what's going on.

clumens commented:
The `Retry sending in crm_ipcs_flush_events` patch fixes the problem in the Remote* tests, but there's still something going on causing other failures:

```
Apr 17 16:40:58 Running test SimulStop              (rhel9-ctslab-3) [  6]
Apr 17 16:46:44 Warn: All nodes stopped but CTS didn't detect: ['rhel9-ctslab-2\\W.*Unloading all Corosync service engines']
Apr 17 16:46:44 Test SimulStopLite                  FAILED: Missing log message: ['rhel9-ctslab-2\\W.*Unloading all Corosync service engines']
Apr 17 16:46:44 Test SimulStop                      FAILED: Stopall failed
```

```
Apr 17 16:59:48 Running test Reattach               (rhel9-ctslab-1) [ 14]
Apr 17 17:02:10 ResourceActivity: 2025-04-17T17:02:04-0400 rhel9-ctslab-1 pacemaker-controld[16925]:  notice: Result of start operation for stateful-1 on rhel9-ctslab-1: OK
Apr 17 17:03:15 Test Reattach                       FAILED: Resources stopped or started after resource management was re-enabled
```

```
Apr 17 17:06:29 Running test NearQuorumPoint        (rhel9-ctslab-3) [ 16]
Apr 17 17:07:10 BadNews: 2025-04-17T17:06:37-0400 rhel9-ctslab-1 pacemaker-controld[27993]:  error: Could not contact the scheduler: Resource temporarily unavailable
Apr 17 17:07:10 BadNews: 2025-04-17T17:06:37-0400 rhel9-ctslab-1 pacemaker-controld[27993]:  error: Input I_ERROR received in state S_POLICY_ENGINE from do_pe_invoke_callback
Apr 17 17:07:10 BadNews: 2025-04-17T17:06:37-0400 rhel9-ctslab-1 pacemaker-controld[27993]:  error: Input I_TERMINATE received in state S_RECOVERY from do_recover
Apr 17 17:07:10 BadNews: 2025-04-17T17:06:37-0400 rhel9-ctslab-1 pacemaker-controld[27993]:  error: 7 resources were active at shutdown
Apr 17 17:07:10 BadNews: 2025-04-17T17:06:37-0400 rhel9-ctslab-1 pacemaker-controld[27993]:  error: Could not recover from internal error
Apr 17 17:07:10 BadNews: 2025-04-17T17:06:37-0400 rhel9-ctslab-1 pacemakerd[27986]:  error: pacemaker-controld[27993] exited with status 1 (Error occurred)
```

These test failures appear to be somewhat transient.  I can't reliably reproduce them on every test every time, but they will happen fairly frequently.  It's difficult to tell from looking at the cts-lab logs which failures represent real errors and which are merely due to the lab killing things and restarting VMs.

clumens commented:
> These test failures appear to be somewhat transient. I can't reliably reproduce them on every test every time, but they will happen fairly frequently. It's difficult to tell from looking at the cts-lab logs which failures represent real errors and which are merely due to the lab killing things and restarting VMs.

This was fixed by the following:
```
-        if (pcmk_all_flags_set(flags, crm_ipc_multipart)) {
-            bool is_end = pcmk_all_flags_set(flags, crm_ipc_multipart_end);
+        if (pcmk__ipc_msg_is_multipart(header)) {
+            bool is_end = pcmk__ipc_msg_is_multipart_end(header);
```

clumens commented:
Rebased onto main to reflect that I merged the compression PR, no other code changes.

nrwahl2 reviewed (commented):
VERY partial. Just getting these early comments out while I continue reviewing.

nrwahl2 commented on lib/common/ipc_client.c:
```diff
@@ -1063,29 +1060,31 @@ long
 
-    client->buffer[0] = 0;
-    client->msg_size = qb_ipcc_event_recv(client->ipc, client->buffer,
+    buffer = pcmk__assert_alloc(crm_ipc_default_buffer_size(), sizeof(char));
```
Another thought: I'm not especially concerned about dynamic memory allocation, but we could avoid it here by using an internal macro constant for buffer size instead of calling `crm_ipc_default_buffer_size()` in these places. Then we could just have a static array for the buffer.

Up to you. I see the appeal of wanting to use that function consistently even internally, but the buffer size is always the same now.

clumens replied:
Even in my followup PR for non-events, all GByteArray manipulations (aside from freeing) are confined to `pcmk__ipc_msg_append`, so once you're reviewed the patch that adds that function, you've seen everything.  We definitely will append without truncating.  That's what will happen in every case aside from the first part of a multi-part message and the case where the entire message fits in one buffer.

I'll also note that it's important to get this right, and that all this IPC code in general is heavily used and therefore timing sensitive.  Anything we can do to make processing faster will help us avoid problems.

clumens commented on lib/common/ipc_server.c:
```diff
@@ -498,7 +498,10 @@ crm_ipcs_flush_events(pcmk__client_t *c)
-        qb_rc = qb_ipcs_event_sendv(c->ipcs, event, 2);
+        do {
+            qb_rc = qb_ipcs_event_sendv(c->ipcs, event, 2);
+        } while (qb_rc == -EAGAIN);
```
https://projects.clusterlabs.org/T322 is relevant here.

nrwahl2 commented on include/crm/common/ipc.h:
```diff
@@ -150,6 +150,10 @@ enum crm_ipc_flags
     //! All replies to proxied connections are sent as events.  This flag
     //! preserves whether the events should be treated as an Event or a Response
     crm_ipc_proxied_relay_response  = (UINT32_C(1) << 18),
+    //! This is a multi-part IPC message
```
All of these internal ones could have doxygen like 
```
/*!
  * \internal
  * \brief [Description]
  */
```

if we care enough. The "internal use only" comment isn't going to show up in the generated docs currently. But it doesn't really matter.

(There also may be a way to do it for all of them with a group directive.)

clumens replied:
We can address this in a separate PR later, since we need to deal the already existing ones too.

nrwahl2 commented on lib/common/ipc_common.c:
```diff
@@ -70,3 +70,39 @@ pcmk__client_type_str(uint64_t client_type)
+        return false;
+    }
+
+    return pcmk_all_flags_set(header->flags, crm_ipc_multipart);
```
`pcmk_is_set()` would be the more intuitive call here and below. They do exactly the same thing, `pcmk_is_set()` is just a wrapper for clarity with a single flag.

And could just be
```
return pcmk__valid_ipc_header(header)
       && pcmk_is_set(header->flags, crm_ipc_multipart);
```

Whatever you think is more readable.

clumens replied:
Having written the code now, I like moving the header check to the callers.  We were actually already doing that everywhere except for one spot, so the checks were redundant.

nrwahl2 commented on lib/common/crmcommon_private.h:
```diff
@@ -314,6 +314,7 @@ typedef struct pcmk__ipc_header_s {
     uint32_t size;
     uint32_t flags;
     uint8_t version;
+    uint16_t part_id;               // If this is a multipart message, which part is this?
```
Nothing wrong with this and it's probably plenty big, but we don't use `uint16_t` anywhere else. Any reason? Just curious.

clumens replied:
I *think* the idea was that a GByteArray's maximum size is `guint`, which is the same as `unsigned int`, which is at least 16 bits, hence `uint16_t`.

nrwahl2 commented on lib/common/ipc_common.c:
```diff
@@ -70,3 +70,39 @@ pcmk__client_type_str(uint64_t client_type)
+    pcmk__ipc_header_t *header = data;
+
+    if (!pcmk__valid_ipc_header(header)) {
+        return false;
```
This function returns a `uint16_t`, not a `bool`. We should either return a magic number (and ideally all these should have trivial doxygen), or return a standard Pacemaker return code with an output argument.

nrwahl2 commented on lib/common/ipc_server.c:
```diff
@@ -649,6 +649,14 @@ pcmk__ipc_prepare_iov(uint32_t request, const GString *message, uint16_t index,
     header->qb.size = iov[0].iov_len + iov[1].iov_len;
     header->qb.id = (int32_t)request;    /* Replying to a specific request */
 
+    if (rc == pcmk_rc_ok && index != 0) {
```
parentheses (style)

nrwahl2 reviewed (commented):
Another partial review

clumens commented on lib/common/ipc_client.c:
```diff
@@ -1246,18 +1263,29 @@ crm_ipc_send(crm_ipc_t *client, const xmlNode *message,
         if (qb_rc < 0) {
             crm_warn("Sending %s IPC disabled until pending reply received",
                      client->server_name);
+            free(buffer);
```
We actually use `buffer` in the else case just below here to check if it's a multipart message or not in order to know if we need to keep looping.

clumens replied:
Since I need to go back to this commit anyway, I've corrected it to use one free.

clumens commented on lib/common/ipc_client.c:
```diff
@@ -1095,17 +1094,23 @@ crm_ipc_read(crm_ipc_t * client)
     if (header) {
-        /* Data excluding the header */
+        client->buffer = g_byte_array_sized_new(client->buf_size);
+        g_byte_array_append(client->buffer, (const guint8 *) buffer,
```
Hm, yes.  I think I misread the GByteArray docs.  It looks like `g_byte_array_append` works by calling memcpy, so I do need to double check callers to make sure they free the buffer if appropriate.

clumens commented:
I've updated this in various ways that make it worth looking at from the beginning again:
* Rebased on main, unfortunately.
* Added a new patch that adds a `pcmk_rc_ipc_more` return code, stopped using `EAGAIN` except to interface with libqb throughout.
* Replaced the patch that looped on `EAGAIN` with one that handles that return code differently for server events.

I have one concern here that comes up occassionally in testing.  I've been setting the buffer size to something small, like 16k, to ensure that messages get split up so I can test all the splitting/combining code.  This results in a lot more IPC traffic, which results in a lot more EAGAIN return codes, and also sometimes means things don't work - for instance, remote nodes sometimes get hung when you try to run `cibadmin -Q` on them.

On the one hand, this is kind of an invented error.  The buffer size is never going to be that small.  I've created this problem by making it easier to test.  On the other hand, I'm concerned that with a very large cluster, we're basically going to see the same problem where there's a lot of IPC traffic, but just with large messages instead of small ones.

clumens commented:
@nrwahl2 I think it's worth closing this PR and focusing on the much smaller #3877, which seems to be working pretty well.  Your comments around GByteArray usage will be applicable to that PR as well, so we could either keep this one open until we hash those out, or we could start over on the new PR - whatever works best for you.

clumens commented on lib/common/ipc_client.c:
```diff
@@ -1161,17 +1168,25 @@ internal_ipc_get_reply(crm_ipc_t *client, int request_id, int ms_timeout,
+                g_byte_array_free(client->buffer, TRUE);
+            }
+
+            client->buffer = g_byte_array_sized_new(client->buf_size);
```
Maybe it's less confusing to just ditch this comment, then.  I like the simplicity of allocating and freeing a buffer for each message.  One thing that might be even nicer to do is allocate the buffer in `crm_ipc_new`, but I am unclear on whether each IPC message is a different connection/client object or not.
